### ISSUE TITLE
[Meta] Up-for-grabs label

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2255]
tl;dr I'd like to apply an <kbd>up-for-grabs</kbd> label as liberally as possible because it's a good signal to potential outside contributors that we are hoping they jump in.

Even if most of our unassigned backlog items end up with this, it's still a valuable signal. Outsiders won't know what our plans mean for them when we leave an issue unassigned, and most can't see the fact that it's in our ZenHub backlog pipeline or what that means for them either.

Right now I'm thinking the ideal process (after awaiting a team decision) would end with either:

 * Closing the issue
 * Assigning to someone
 * Putting it in the ZenHub backlog pipeline and labeling up-for-grabs unless there is a reason not to (see **What don't we label as up-for-grabs?**)

I'd like it in green because psychology.

### Isn't everything up for grabs?

 1. I _don't_ think we should take the position that labeling up-for-grabs will have no effect. It may nudge people that would have been on the fence, if only to communicate our culture to them. It may influence their mindset if they don't think about open source they way we wish they did. We should attract people so that they know we want them and they don't have to guess. Let's be as communicative as we can.

    It's also a way to tell people that if they want to see a certain thing finished sooner, it'll likely happen soonest if they help. Knowing that we're planning for outside contributors to work on an issue will itself make it more likely for outside contributors to work on it.

 2. This would also open the door to https://github.com/up-for-grabs/up-for-grabs.net#add-a-project, which I am hopeful for.

 3. Having the label would also make it more easy and discoverable for a potential contributor to browse the list of up-for-grabs issues. Right now all they'll see if they go looking is... so many issues.

    \<hypothetical>
    It would also work well with a <kbd>wishlist</kbd> label. ([Corefx example](https://github.com/dotnet/corefx/labels/wishlist)) for occasional issues that are really our favorite that we don't have the luxury to spend time on, but would be amazing if someone dropped by and it caught their eye out of the list.
    \</hypothetical>

 4. Applying the label should require us to clearly describe the work and leave pointers to help the implementer before tossing it on the backlog. This makes it more likely that an outsider will decide to help, which we could use. It'll even be helping ourselves down the road, especially if our team members change.

### What don't we label as up-for-grabs?

If we get visibility through http://up-for-grabs.net, they recommend using the label for these types of issues:

 * Tasks should take no longer than a few nights' worth of work
 * Tasks should stand alone - avoid core functionality on which other tasks depend
 * Tasks should be well described with pointers to help the implementer

That makes sense.

/cc @nunit/framework-team I'm hoping for input from everyone. Would you be open to trying this out?

**Comment by ChrisMaddock** [comment]
I suggested this before, but it was decided against on the basis that really, everything is 'up-for-grabs'. 

Now, I'm not so sure on it. I like the idea, but it is a valid point that we're not a professional team, so really anything unassigned should be open for community contribution. The only state that might prevent that, would be that we haven't reached a decision on something yet, which does cover a fair few open issues.

**Comment by jnm2** [comment]
Cool. I organized the sections above a little better. How would you respond to the points above?

**Comment by ChrisMaddock** [comment]
I they're are all good points, although I don't totally understand what you're getting at with the first sentence of point 1.

It comes back to the same issue as when I suggested it however long ago though - what _don't_ we label as up-for-grabs? (That may actually be a better point for you to clarify on! 🙂 )

**Comment by CharliePoole** [comment]
Our closest label right now is easyfix, but that adds the qualification that the job is easy.

One thing that @jnm2 brings up has been bothering me too. In order to understand what it's OK to work on you have to have ZenHub. That's not helpful.

I didn't know about up-for-grabs.net. That's at least a small motivator toward having a label. However, I'm inclined to think that we would have to do some work on spelling out how we want something to be done if we marked it up-for-grabs - i.e. some bit of up-front design, which we usually don't do.

**Comment by jnm2** [comment]
@ChrisMaddock 
> It comes back to the same issue as when I suggested it however long ago though - what don't we label as up-for-grabs?

@CharliePoole 
> However, I'm inclined to think that we would have to do some work on spelling out how we want something to be done if we marked it up-for-grabs - i.e. some bit of up-front design, which we usually don't do.

Appreciated! I addressed both of these in point 4 and added a section "What don't we label as up-for-grabs?"

**Comment by jnm2** [comment]
It's also a way to tell people, "if you want this, don't wait for us to do it." Knowing that we're planning for outside contributors to work on an issue will itself make it more likely for outside contributors to work on it.

**Comment by CharliePoole** [comment]
@jnm2 The points you list under when not to use it map fairly well to the criteria we use for easyfix. Since up-for-grabs.net allows you to specify the label used in your yml file, it seems like we don't need another label. However, we may need to document better when we want easyfix to be applied and/or follow those guidelines better.

**Comment by CharliePoole** [comment]
@jnm2 On your last comment, I disagree. I do not think we want up-for-grabs or easyfix to mean we won't do it but you can or to have anything to do with priorities. It's orthogonal. If you want to indicate we probably won't get to it, just label it pri:low.

**Comment by jnm2** [comment]
@CharliePoole
>The points you list under when not to use it map fairly well to the criteria we use for easyfix.

The only reservation I have is that I really want Randomizer.NextGuid() to be something that is marked up-for-grabs, but you didn't think it was an easyfix.  It's not the easiest thing in the world, but it's a self-contained, documented issue that anyone could pick up if they wanted.

> I do not think we want up-for-grabs or easyfix to mean we won't do it but you can or to have anything to do with priorities.

Fair. How about, "It's also a way to tell people that if they want to see a certain thing finished sooner, it'll likely happen soonest if they help."

**Comment by CharliePoole** [comment]
I guess I figured one of our existing commiters or contributors should do this since Randomizer is so central to the framework. We could add some guidance that would steer people away from breaking test execution in general.

**Comment by CharliePoole** [comment]
One reason I am hesitant about solving this with a label is that I want to see us add comments that say "If you pick this up, do this... don't do that..."

**Comment by ChrisMaddock** [comment]
I definitely think there's two different concepts in 'easyfix' and 'up-for-grabs'. The former should be for simple changes suitable for a first time contributor to open source in general. The latter can have a bit more meat to it.

@jnm2 - I think your points under "What don't we label as up-for-grabs?" are good. I think the scoping is key. As you said, @CharliePoole - issues would need to have a certain amount of design spec done before adding an 'up-for-grabs'.

I like the idea. 👍

**Comment by jnm2** [comment]
@CharliePoole
> I guess I figured one of our existing commiters or contributors should do this since Randomizer is so central to the framework. We could add some guidance that would steer people away from breaking test execution in general.

Is that sustainable though? And don't PRs reviews deal with this well enough? I welcome outside implementations that break things because it means we don't have enough tests. And where there are tests, that's the best spec and tightest feedback loop possible for the outside contributor to develop against.

> One reason I am hesitant about solving this with a label is that I want to see us add comments that say "If you pick this up, do this... don't do that..."

I think we need both. The label for discoverability and advertising if you will, and a rule that you don't apply the label without a clear description with pointers.

@ChrisMaddock 
> I definitely think there's two different concepts in 'easyfix' and 'up-for-grabs'. The former should be for simple changes suitable for a first time contributor to open source in general. The latter can have a bit more meat to it.

That's a great way to put it!

**Comment by jnm2** [comment]
Keeping `easyfix` alongside `up-for-grabs` is also a really good idea because outside contributors will notice what is easy and what is more meaty.

**Comment by CharliePoole** [comment]
@jnm2 I think finding out you did the wrong thing only after you write the code is a bad experience for a new contributor.

**Comment by jnm2** [comment]
@CharliePoole
>  I think finding out you did the wrong thing only after you write the code is a bad experience for a new contributor.

I don't disagree, but it's also a reality for every programmer. Many of my PRs to other projects seemed to tend that direction. I'm going to borrow your point that programmers are a resourceful lot. 😀 If we have so few tests that this ends up happening all the time, we'll recognize that and take the initiative to write more internal tests.
We can make it a good experience for new and experienced contributors by expressing that each submission is tremendously appreciated by us whether it reveals tests we overlooked writing or goes to production. If we apologize for overlooking an important aspect and thank them for uncovering it, won't they be sympathetic?

More importantly, I don't see a way around it if our goal is to get as many people as possible working on NUnit. Which it is... right?

**Comment by CharliePoole** [comment]
I think it's one thing to get your code sent back as a result of code review. It's another to discover you did something completely different from what you were supposed to do. There may not even be test failures, but the reviewers are saying... that's not what we wanted from this issue. We should avoid that.

With failing tests, we avoid the problem by having better tests. With doing-the-wrong-thing errors we do it by having better issues.

Here's our existing definition of easyfix in the docs...

```
**easyfix** Indicates an issue that might be a good place for a new contributor to start. Whoever adds the label should couple it with a comment suggesting what code to look at and a general approach to working the issue.
```

In writing it, what I had in mind was new contributors __to NUnit__, not necessarily new to Open Source, and I assume that they are experienced programmers to start with.

What would you change? How would you define up-for-grabs?

**Comment by jnm2** [comment]
@CharliePoole Okay, yes. I agree.

> What would you change? How would you define up-for-grabs?

First attempt:

<table><tr><td>

##### `up-for-grabs`

Indicates a backlog issue ready for implementation where we would love for the wider community to jump in to be assigned! Whoever adds this label should couple it with a comment suggesting what code to look at and a general approach to working the issue.

##### `easyfix`

Indicates an issue that is expected to take no more than a few nights' work and be accessible to new contributors.
</td></tr></table>

**Comment by CharliePoole** [comment]
I like your definition of `up-for-grabs`. It makes me think of just renaming it "community".

Isn't "Whoever..." still needed on `easyfix`? Seems like the new folks need the guidance even more. As a possible separate matter, we once considered assigning mentors to new contributors. Of course, we do that informally anyway.

**Comment by jnm2** [comment]
> It makes me think of just renaming it "community".

I'd prefer that except that `up-for-grabs` is possibly more immediately recognizable?

> Isn't "Whoever..." still needed on easyfix?

We should either require `easyfix` to be accompanied with `up-for-grabs`, or we should purposely allow it to be used on things that are *not* up for grabs. I could see potential value in either direction.

**Comment by rprouse** [comment]
I did a PR on Up for Grabs a couple years ago to add our easyfix label, so that is what we are currently listed under. I'm not sure it ever brought in many people, but who knows 😄 

I like `up-for-grabs`. It might encourage people to contribute rather than repeatedly ask "what is the status of this issue?" or "when are you going to do this?"

If we do switch, we should make the change on Up for Grabs too.

**Comment by jnm2** [comment]
> I did a PR on Up for Grabs a couple years ago to add our easyfix label, so that is what we are currently listed under. I'm not sure it ever brought in many people, but who knows 😄

So we are! 😆 I didn't look hard enough.
We are under http://up-for-grabs.net/#/tags/testing but sadly not under [http://up-for-grabs.net/#/tags/unit testing](http://up-for-grabs.net/#/tags/unit%20testing). I wonder if that can be fixed?

**Comment by jnm2** [comment]
Other possibilities:
 - http://www.firsttimersonly.com/#are-you-an-oss-project-owner
 - https://www.codetriage.com/?language=C%23

**Comment by CharliePoole** [comment]
Funny! Which of our projects are listed?

The tags would be in the yml file that Rob submitted.

**Comment by CharliePoole** [comment]
I guess we are overthinking this. Let's either keep easyfix or change it to up-for-grabs or help-wanted or whatever we like best - but one label seems sufficient.

**Comment by jnm2** [comment]
@CharliePoole I want a label I can put on noteasyfixes so... 😆

**Comment by jnm2** [comment]
https://github.com/up-for-grabs/up-for-grabs.net/blob/gh-pages/_data/projects/nunit.yml

**Comment by rprouse** [comment]
I think we are overthinking this too. Let's just add an `up-for-grabs` label and see how it works 😄

**Comment by jnm2** [comment]
There's no way back from deleting `easyfix` if we change our minds so let's keep it. 😈

**Comment by jnm2** [comment]
Also, we have `hardfix` and that makes it a good symmetry to leave `easyfix`.

My feeling is that the smallest possible change is to leave `easyfix` labels in place and add an `up-for-grabs` label. Is that a good way to proceed?

**Comment by ChrisMaddock** [comment]
I agree with keeping easyfix alongside up-for-grabs. I think it's good to have a tag to identify the simplest issues to newcomers to pick up, and it seems we want to encompass more than that in up-for-grabs.

**Comment by jnm2** [comment]
Another reason to keep `easyfix` is that we'd want to keep the easy vs hard classification while removing `up-for-grabs` when the issue is no longer up for grabs.

A common scenario will be first-time contributors. I've picked up on a preference for us not to add them as collaborators until we've seen a few contributions from them. That means we will be leaving the issue unassigned. Removing `up-for-grabs` will be the best way to signal that the issue is taken.

**Comment by jnm2** [comment]
It looks like we have consensus on creating an `up-for-grabs` label and updating the descriptions as in
https://github.com/nunit/nunit/issues/2255#issuecomment-308846068.

As we have time, we should wander through all unassigned backlog items and add or get the ball rolling on necessary documentation to be able to add an `up-for-grabs` tag.

**Comment by jnm2** [comment]
We should start with `easyfix` issues and once they are all documented (if perchance one isn't already) and have `up-for-grabs` applied, we should update our up-for-grabs.net tag.

**Comment by jnm2** [comment]
I've reviewed the few `easyfix` issues we have, and I commented on the only two issues remaining which might need tags updated. After they are updated, we can switch our up-for-grabs.net tag.

Next, unassigned backlog items that aren't necessarily easyfix: [is:open no:assignee -label:up-for-grabs](https://github.com/nunit/nunit/issues?utf8=%E2%9C%93&q=is%3Aopen%20no%3Aassignee%20-label%3Aup-for-grabs%20)

**Comment by CharliePoole** [comment]
@jnm2 I thought we had a consensus, so I don't want to restart that discussion, at least.

However, your last comment about examining unassigned backlog items skips the triage process. We have plenty of items hat are open unassigned issues because we have not yet decided whether to do them. For that you have to look at ZenHub as well. 

[I'm not as enthused about ZenHub as I once was, but since we are using it we have to look at the pipeline assigned to each one.]

Bottom line, I think there has to be an explicit decision that we want to do something before it can either be self-assigned, labelled up-for-grabs or have a PR accepted on it.

**Comment by jnm2** [comment]
@CharliePoole Agreed. Thanks for bringing that up. I was thinking we should triage them and ask for a decision with an eye to either closing or marking `up-for-grabs`.

**Comment by jnm2** [comment]
See https://github.com/up-for-grabs/up-for-grabs.net/pull/654

**Comment by rprouse** [comment]
I think this is done. Reopen if I am mistaken.
